tools/autograd/gen_autograd_functions.py

```diff
@@ -291,19 +291,21 @@ def process_function(info: DifferentiabilityInfo, template: CodeTemplate) -> str
         compute_index_ranges.append(f'auto {arg.name}_ix = gen.range({size});')
 
     def save_var(var: SavedAttribute, is_output: bool) -> None:
-        name = var.name
+        name = var.nctype.name
+        type = var.nctype.type
         should_append_getsetdef = True
 
-        if var.type == 'at::Tensor' or var.type == 'c10::optional<at::Tensor>' or var.type == 'c10::optional<at::Tensor>&' or \
-                (var.type == 'at::Scalar' and is_output):
+        if type == BaseCType(tensorT) or type == OptionalCType(BaseCType(tensorT)) or \
```

Comment:
maybe we should have some sort of set of `tensorBT` names for `BaseCType(tensorT)` ha!